Adding the Actor Migration Kit.

vjovanov opened the pull request:
Kit consists of:
    1) The StashingActor which adopts an interface similar to Akka.
    2) Props mockup for creating Akka like code
    3) Pattern mockup
    4) Test cases for every step in the migration.
    5) MigrationSystem which will paired on the Akka side.

  Review of the code : @phaller
  Review of the build: @jsuereth (scala-actors-migration.jar is the new jar containing the migration code)

Commit:
Adding the Actor Migration Kit.

  Kit consists of:
    1) The StashingActor which adopts an interface similar to Akka.
    2) Props mockup for creating Akka like code
    3) Pattern mockup
    4) Test cases for every step in the migration.
    5) MigrationSystem which will paired on the Akka side.

  Review of the code : @phaller
  Review of the build: @jsuereth

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/122/

jsuereth commented on src/partest/scala/tools/partest/nest/SBTRunner.scala:
```diff
@@ -65,6 +66,7 @@ object SBTRunner extends DirectRunner {
     fileManager.LATEST_COMP = findClasspath("scala-compiler", "scala-compiler") getOrElse sys.error("No scala-compiler found! Classpath = " + fileManager.CLASSPATH)
     fileManager.LATEST_PARTEST = findClasspath("scala-partest", "partest") getOrElse sys.error("No scala-partest found! Classpath = " + fileManager.CLASSPATH)
     fileManager.LATEST_ACTORS = findClasspath("scala-actors", "actors") getOrElse sys.error("No scala-actors found! Classpath = " + fileManager.CLASSPATH)
+    fileManager.LATEST_ACTORS_MIGRATION = findClasspath("scala-actors-migration", "actors-migration") getOrElse sys.error("No scala-actors-migration found! Classpath = " + fileManager.CLASSPATH)
```
Thanks for keeping SBT working!

jsuereth commented:
build LGTM, although we need some tweaks to scala-dist most likely after this goes in.   Can you open a ticket for me?

https://github.com/scala/scala-dist/issues/new

Thanks!

phaller commented on src/actors-migration/scala/actors/Pattern.scala:
```diff
@@ -0,0 +1,25 @@
+
+  def !(message: Any)(implicit sender: ActorRef = null): Unit = ar.!(message)(sender)
+
+  def ?(message: Any)(timeout: Timeout): Future[Any] = ar.?(message, timeout.duration)
```
Maybe add a doc comment?

vjovanov replied:
Yes I will. Documentation will come in the RC phase.

phaller replied:
OK, sounds good.

phaller commented on src/actors-migration/scala/actors/Props.scala:
```diff
@@ -0,0 +1,14 @@
+
+  /**
+   * Returns a new Props with the specified creator set
+   *  Scala API
```
I'd probably remove the "Scala API" note. :-)

vjovanov replied:
Fixed.

phaller commented on src/actors-migration/scala/actors/StashingActor.scala:
```diff
@@ -0,0 +1,261 @@
+  }
+}
+
+@deprecated("Scala Actors are beeing removed from the standard library. Please refer to the migration guide.", "2.10")
```
beeing -> being

vjovanov replied:
Fixed.

phaller commented on src/actors-migration/scala/actors/StashingActor.scala:
```diff
@@ -0,0 +1,261 @@
+  def unhandled(message: Any) {
+    println("unhandeld")
+    message match {
+      case _ => throw new UnhandledMessageException(message, self)
```
I think we shouldn't throw that exception by default. At least in Akka 2.0 there is no exception being thrown by default.

vjovanov replied:
Good catch, It must date back before Akka 2.0. Will fix it.

vjovanov replied:
In Akka they now use an event stream for unhandeled messages. What do you say we use the System.err? 
I think that it is an OK alternative for the migration purposes only.

phaller commented on src/actors-migration/scala/actors/StashingActor.scala:
```diff
@@ -0,0 +1,261 @@
+/**
+ * This message is thrown by default when an Actors behavior doesn't match a message
+ */
+case class UnhandledMessageException(msg: Any, ref: ActorRef = null) extends Exception {
```
We should probably not throw such exceptions by default (see above), but maybe users might want to chose to throw such an exception. On the other hand, if we aren't using the exception, then it might be best to just drop that class.

vjovanov replied:
FIxed.

phaller commented on src/actors-migration/scala/actors/Timeout.scala:
```diff
@@ -0,0 +1,34 @@
+/**
+ * Copyright (C) 2009-2011 Typesafe Inc. <http://www.typesafe.com>
```
Use the regular Scala header instead. Same for the other files.

vjovanov replied:
Fixed.

phaller commented on src/actors/scala/actors/ActorRef.scala:
```diff
@@ -0,0 +1,119 @@
+  /**
+   * Sends a message asynchronously, returning a future which may eventually hold the reply.
+   */
+  private[actors] def ?(message: Any, timeout: Duration): Future[Any]
```
Isn't that method added using the ask support? If so, we should remove the method here.

vjovanov replied:
It is ask support internals. That is why it is private[actors].

phaller commented on test/files/jvm/actmig-PinS.scala:
```diff
@@ -0,0 +1,118 @@
+import scala.actors._
```
Awesome job on all the tests, btw :-)

vjovanov replied:
Thanks

Commit:
Minor fixes for the Actor Migration Kit

  Fixed behavior of the StashingActor in case of unhandeled message.
  Fixed a typo in deprecated annotation.
  Fixed comments.
  Fixed copyright.

adriaanm commented:
Cool! So, @vjovanov, is the pull request good to go? I assume the "fixed" comments relate to the commit you added 2 hours ago?

vjovanov commented:
Yes that request is the fix for all the comments. It is good to go.

adriaanm commented:
the build failed when packing: https://scala-webapps.epfl.ch/jenkins/job/scala-checkin/5746/consoleText

jsuereth commented:
AHHH, you missed the tricksy "scaladocs are copied for standard library so we have one for each artifact to satisfy sonatype" madness.  So did I apparently.    Want me to patch it?

vjovanov commented:
In this case I generate scala-doc because the migration kit is completely separate than the lib. It is not even in the library folder. Also I did not completely miss it but I used a wrong path. I will patch it, thanks for the offer :)

adriaanm commented:
thanks -- it becomes harder to judge new pull requests when master does not pass the test suite, so a quick resolution would be great!

vjovanov commented:
I did not understand that the master is broken. Since I squashed the actor migration commits can you revert the master and I will submit a new pull request. It is currently building on the jenkins app.
